### ISSUE TITLE
Webhook payload does not conform to NetBox OpenAPI schema

### ISSUE THREAD

**Issue by commonism** [peteeckel/netbox-plugin-dns#39]
First of all - thanks for taking over this project.

**Versions**
NetBox Version: netbox-docker
NetBox DNS Version: pypi
Python Version: netbox-docker

**Describe the bug**
Using netbox webhooks …

Usually the format of the payload you get is 
```json
{
    "event": "updated",
    "timestamp": "2023-08-01 11:57:37.987666+00:00",
    "model": "zone",
    "username": "admin",
    "request_id": "36bc8845-03c8-490b-a466-552b9d10a77e",
    "data": <…>
    "snapshots": {
        "prechange": {
            <…Request>
        },
        "postchange": {
            <…Request>
    }
}
```

You can lookup the format of <…> and <…> from the openapi.yaml and have some python glue do the magic to create models, inject these into fastapi, and voiola, there is your webhook with perfect validation & parsing of the data, using models instead of basic python data structures.
A project I'm working on "netbox-webhook`, meant to use netbox-dns webhooks to interface a name using TSIG.

Currently the format of the webhook payload data of netbox-dns does not match the <…>Request JSON schemas from the description document.

For one, some fields are missing allow_null

```diff
--- serializers.py.1	2023-08-01 14:24:37.138861089 +0200
+++ serializers.py	2023-08-01 13:33:01.938615897 +0200
@@ -54,6 +54,7 @@
     active = serializers.BooleanField(
         required=False,
         read_only=True,
+        allow_null=True,
     )
 
     def create(self, validated_data):
@@ -141,6 +142,7 @@
         many=False,
         read_only=True,
         required=False,
+        allow_null=True,        
         help_text="PTR record generated from an address",
     )
     address_record = NestedRecordSerializer(
@@ -157,6 +159,7 @@
     active = serializers.BooleanField(
         required=False,
         read_only=True,
+        allow_null=True,        
     )
 
     class Meta:
```

Next, …
```python
fastapi.exceptions.RequestValidationError: [
{'type': 'model_attributes_type', 'loc': ('body', 'zone', 'snapshots', 'prechange', 'view'), 'msg': 'Input should be a valid dictionary or object to extract fields from', 'input': 1, 'url': 'https://errors.pydantic.dev/2.0.3/v/model_attributes_type'}, 
{'type': 'model_attributes_type', 'loc': ('body', 'zone', 'snapshots', 'prechange', 'nameservers', 0), 'msg': 'Input should be a valid dictionary or object to extract fields from', 'input': 4, 'url': 'https://errors.pydantic.dev/2.0.3/v/model_attributes_type'}, 
{'type': 'model_attributes_type', 'loc': ('body', 'zone', 'snapshots', 'prechange', 'soa_mname'), 'msg': 'Input should be a valid dictionary or object to extract fields from', 'input': 4, 'url': 'https://errors.pydantic.dev/2.0.3/v/model_attributes_type'}, 
…
```

The Payload looks like this:

```yaml
{
    "event": "updated",
    "timestamp": "2023-08-01 12:26:09.992279+00:00",
    "model": "zone",
    …
    },
    "snapshots": {
        "prechange": {
            "created": "2023-08-01T11:00:22.010Z",
            "last_updated": "2023-08-01T11:57:37.949Z",
            "view": 1,
            "name": "example.org",
            "status": "active",
            "default_ttl": 86400,
            "soa_ttl": 86400,
            "soa_mname": 4,
            "soa_rname": "hostmaster.example.org",
            "soa_serial": 1,
            "soa_refresh": 172800,
            "soa_retry": 7200,
            "soa_expire": 2592000,
            "soa_minimum": 3600,
            "soa_serial_auto": false,
            "description": "",
            "arpa_network": null,
            "nameservers": [
                4
            ],
            "custom_fields": {},
            "tags": []
        },
…
```

The Schema for ZoneRequest - retrieved via /api/schema :
```yaml
    ZoneRequest:
      type: object
      description: Adds support for custom fields and tags.
      properties:
        name:
          type: string
          minLength: 1
          maxLength: 255
        view:
          allOf:
          - $ref: '#/components/schemas/NestedViewRequest'
          description: View the zone belongs to
        nameservers:
          type: array
          items:
            $ref: '#/components/schemas/NestedNameServerRequest'
          description: Nameservers for the zone
…
```

I've found WritableZoneRequest - which does better wrt. to views, but fails for nameservers as well.
And there is no WritableViewRequest either.

```yaml
    WritableZoneRequest:
      type: object
      description: Adds support for custom fields and tags.
      properties:
        name:
          type: string
          minLength: 1
          maxLength: 255
        view:
          type: integer
          nullable: true
        nameservers:
          type: array
          items:
            $ref: '#/components/schemas/NestedNameServerRequest'
          description: Nameservers for the zone
```

I lack the insight in django serialization to get this done by myself.

I currently hook ipam.IPAddress for dns_name and would like to extend to netbox-dns.

**Comment by peteeckel** [comment]
Hi @commonism, thanks for raising this!

I've created a separate issue for the `allow_null` stuff. Though I can't currently reproduce this either, it doesn't seem to hurt or break any tests so I'll add the attribute where you suggested it. If I may ask, please don't include multiple problems in one issue if possible but rather open a single issue for each, that makes it much easier to track.

Regarding the OpenAPI specs returned by the `/api/schema` endpoint: These are automatically generated by Django/DRF (Django REST Framework). There have been some problems with the automatically generated API documentation in the past that I'm aware of, but I never figured out how to convince the OpenAPI/Swagger toolchain to return the correct results - some were quietly fixed by one of the last NetBox/DRF upgrades.

The `RequestValidationError` created by FastAPI looks like the webhook code passes the `id` of the nameserver as an element of `nameservers` and as `soa_mname`, and the `id` of the view as `view`, while FastAPI, after processing the OpenAPI schema expects a nested object. I'm not sure there even is an option to change what data is passed to the webhook from a plugin - but that seems to be the point.

So it seems that NetBox does not create the payload of the webhook according to the schema it generated automatically. Which is pretty ugly as I don't have any control over either from the plugin code, at least none that I know of. I'll try to find out whether - and if so, how - anything can be done about it but waiting for a DRF upgrade to fix it.

**Comment by commonism** [comment]
Yes, basically the nested objects are serialized to the id integer only - but the schema claims it would be an object.

**Comment by peteeckel** [comment]
Thanks for confirming. Now I need to talk to the NetBox developers ... I hope they have some information about what's happening there.

Can you send me the full payload of the webhook? If necessary, redact out host names but keep the structure complete.

**Comment by peteeckel** [comment]
Another question: Is it possible that FastAPI is trying to evaluate the snapshots (`prechange`/`postchange`)? Do you actually need the data from the snapshots, or can they be excluded from processing somehow?

**Comment by commonism** [comment]
Complete Payload for zone update:

```json
{
    "event": "updated",
    "timestamp": "2023-08-01 12:26:09.992279+00:00",
    "model": "zone",
    "username": "admin",
    "request_id": "9f38c1c4-f3be-4b5d-ab10-15463cf3dbc7",
    "data": {
        "id": 1,
        "url": "/api/plugins/netbox-dns/zones/1/",
        "name": "example.org",
        "view": {
            "id": 1,
            "url": "/api/plugins/netbox-dns/views/1/",
            "display": "Internal",
            "name": "Internal"
        },
        "display": "[Internal] example.org",
        "nameservers": [
            {
                "id": 4,
                "url": "/api/plugins/netbox-dns/nameservers/4/",
                "display": "ns1.example.org",
                "name": "ns1.example.org"
            }
        ],
        "status": "active",
        "description": "",
        "tags": [],
        "created": "2023-08-01T11:00:22.010262Z",
        "last_updated": "2023-08-01T12:26:09.938379Z",
        "default_ttl": 86400,
        "soa_ttl": 86400,
        "soa_mname": {
            "id": 4,
            "url": "/api/plugins/netbox-dns/nameservers/4/",
            "display": "ns1.example.org",
            "name": "ns1.example.org"
        },
        "soa_rname": "hostmaster.example.org",
        "soa_serial": 1,
        "soa_serial_auto": false,
        "soa_refresh": 172800,
        "soa_retry": 7200,
        "soa_expire": 2592000,
        "soa_minimum": 3600,
        "active": true,
        "custom_fields": {}
    },
    "snapshots": {
        "prechange": {
            "created": "2023-08-01T11:00:22.010Z",
            "last_updated": "2023-08-01T11:57:37.949Z",
            "view": 1,
            "name": "example.org",
            "status": "active",
            "default_ttl": 86400,
            "soa_ttl": 86400,
            "soa_mname": 4,
            "soa_rname": "hostmaster.example.org",
            "soa_serial": 1,
            "soa_refresh": 172800,
            "soa_retry": 7200,
            "soa_expire": 2592000,
            "soa_minimum": 3600,
            "soa_serial_auto": false,
            "description": "",
            "arpa_network": null,
            "nameservers": [
                4
            ],
            "custom_fields": {},
            "tags": []
        },
        "postchange": {
            "created": "2023-08-01T11:00:22.010Z",
            "last_updated": "2023-08-01T12:26:09.938Z",
            "view": 1,
            "name": "example.org",
            "status": "active",
            "default_ttl": 86400,
            "soa_ttl": 86400,
            "soa_mname": 4,
            "soa_rname": "hostmaster.example.org",
            "soa_serial": 1,
            "soa_refresh": 172800,
            "soa_retry": 7200,
            "soa_expire": 2592000,
            "soa_minimum": 3600,
            "soa_serial_auto": false,
            "description": "",
            "arpa_network": null,
            "nameservers": [
                4
            ],
            "custom_fields": {},
            "tags": []
        }
    }
}
```

for Record create
error
```
fastapi.exceptions.RequestValidationError: [
{'type': 'model_attributes_type', 'loc': ('body', 'zone', 'snapshots', 'postchange', 'view'), 'msg': 'Input should be a valid dictionary or object to extract fields from', 'input': 1, 'url': 'https://errors.pydantic.dev/2.0.3/v/model_attributes_type'}, 
{'type': 'model_attributes_type', 'loc': ('body', 'zone', 'snapshots', 'postchange', 'nameservers', 0), 'msg': 'Input should be a valid dictionary or object to extract fields from', 'input': 4, 'url': 'https://errors.pydantic.dev/2.0.3/v/model_attributes_type'}, 
{'type': 'model_attributes_type', 'loc': ('body', 'zone', 'snapshots', 'postchange', 'soa_mname'), 'msg': 'Input should be a valid dictionary or object to extract fields from', 'input': 4, 'url': 'https://errors.pydantic.dev/2.0.3/v/model_attributes_type'}]
```
message
```json
{
    "event": "updated",
    "timestamp": "2023-08-01 18:51:13.038144+00:00",
    "model": "zone",
    "username": "admin",
    "request_id": "58500dcd-a1bc-40e8-8b98-bf7b62eef546",
    "data": {
        "id": 2,
        "url": "/api/plugins/netbox-dns/zones/2/",
        "name": "10.in-addr.arpa",
        "view": {
            "id": 1,
            "url": "/api/plugins/netbox-dns/views/1/",
            "display": "Internal",
            "name": "Internal"
        },
        "display": "[Internal] 10.in-addr.arpa",
        "nameservers": [
            {
                "id": 4,
                "url": "/api/plugins/netbox-dns/nameservers/4/",
                "display": "ns1.example.org",
                "name": "ns1.example.org"
            }
        ],
        "status": "active",
        "description": "",
        "tags": [],
        "created": "2023-08-01T11:26:38.270226Z",
        "last_updated": "2023-08-01T18:51:12.945305Z",
        "default_ttl": 86400,
        "soa_ttl": 86400,
        "soa_mname": {
            "id": 4,
            "url": "/api/plugins/netbox-dns/nameservers/4/",
            "display": "ns1.example.org",
            "name": "ns1.example.org"
        },
        "soa_rname": "hostmaster.example.org",
        "soa_serial": 1690915873,
        "soa_serial_auto": true,
        "soa_refresh": 172800,
        "soa_retry": 7200,
        "soa_expire": 2592000,
        "soa_minimum": 3600,
        "active": null,
        "custom_fields": {}
    },
    "snapshots": {
        "prechange": null,
        "postchange": {
            "created": "2023-08-01T11:26:38.270Z",
            "last_updated": "2023-08-01T18:51:12.945Z",
            "view": 1,
            "name": "10.in-addr.arpa",
            "status": "active",
            "default_ttl": 86400,
            "soa_ttl": 86400,
            "soa_mname": 4,
            "soa_rname": "hostmaster.example.org",
            "soa_serial": 1690915873,
            "soa_refresh": 172800,
            "soa_retry": 7200,
            "soa_expire": 2592000,
            "soa_minimum": 3600,
            "soa_serial_auto": true,
            "description": "",
            "arpa_network": "10.0.0.0/8",
            "nameservers": [
                4
            ],
            "custom_fields": {},
            "tags": []
        }
    }
}
```
the ptr created 
error
```
fastapi.exceptions.RequestValidationError: [
{'type': 'missing', 'loc': ('body', 'record', 'data', 'ptr_record', 'zone', 'active'), 'msg': 'Field required', 'input': {'id': 2, 'url': '/api/plugins/netbox-dns/zones/2/', 'display': '[Internal] 10.in-addr.arpa', 'name': '10.in-addr.arpa', 'view': {'id': 1, 'url': '/api/plugins/netbox-dns/views/1/', 'display': 'Internal', 'name': 'Internal'}, 'status': 'active'}, 'url': 'https://errors.pydantic.dev/2.0.3/v/missing'}, 
{'type': 'missing', 'loc': ('body', 'record', 'data', 'ptr_record', 'active'), 'msg': 'Field required', 'input': {'id': 7, 'url': '/api/plugins/netbox-dns/records/7/', 'display': '2.0.0.10.in-addr.arpa [PTR]', 'type': 'PTR', 'name': '2.0.0', 'value': 'ns2.example.org.example.org.', 'status': 'active', 'ttl': None, 'zone': {'id': 2, 'url': '/api/plugins/netbox-dns/zones/2/', 'display': '[Internal] 10.in-addr.arpa', 'name': '10.in-addr.arpa', 'view': {'id': 1, 'url': '/api/plugins/netbox-dns/views/1/', 'display': 'Internal', 'name': 'Internal'}, 'status': 'active'}}, 'url': 'https://errors.pydantic.dev/2.0.3/v/missing'}, 
{'type': 'model_attributes_type', 'loc': ('body', 'record', 'data', 'address_record'), 'msg': 'Input should be a valid dictionary or object to extract fields from', 'input': None, 'url': 'https://errors.pydantic.dev/2.0.3/v/model_attributes_type'}, 
{'type': 'model_attributes_type', 'loc': ('body', 'record', 'snapshots', 'postchange', 'zone'), 'msg': 'Input should be a valid dictionary or object to extract fields from', 'input': 1, 'url': 'https://errors.pydantic.dev/2.0.3/v/model_attributes_type'}]
```
message
```json

{
    "event": "created",
    "timestamp": "2023-08-01 18:51:13.039416+00:00",
    "model": "record",
    "username": "admin",
    "request_id": "58500dcd-a1bc-40e8-8b98-bf7b62eef546",
    "data": {
        "id": 8,
        "url": "/api/plugins/netbox-dns/records/8/",
        "zone": {
            "id": 1,
            "url": "/api/plugins/netbox-dns/zones/1/",
            "display": "[Internal] example.org",
            "name": "example.org",
            "view": {
                "id": 1,
                "url": "/api/plugins/netbox-dns/views/1/",
                "display": "Internal",
                "name": "Internal"
            },
            "status": "active",
            "active": true
        },
        "display": "ns2.example.org.example.org [A]",
        "type": "A",
        "name": "ns2.example.org",
        "value": "10.0.0.2",
        "status": "active",
        "ttl": null,
        "description": "",
        "tags": [],
        "created": "2023-08-01T18:51:12.982270Z",
        "last_updated": "2023-08-01T18:51:12.982309Z",
        "managed": false,
        "disable_ptr": false,
        "ptr_record": {
            "id": 7,
            "url": "/api/plugins/netbox-dns/records/7/",
            "display": "2.0.0.10.in-addr.arpa [PTR]",
            "type": "PTR",
            "name": "2.0.0",
            "value": "ns2.example.org.example.org.",
            "status": "active",
            "ttl": null,
            "zone": {
                "id": 2,
                "url": "/api/plugins/netbox-dns/zones/2/",
                "display": "[Internal] 10.in-addr.arpa",
                "name": "10.in-addr.arpa",
                "view": {
                    "id": 1,
                    "url": "/api/plugins/netbox-dns/views/1/",
                    "display": "Internal",
                    "name": "Internal"
                },
                "status": "active"
            }
        },
        "address_record": null,
        "active": null,
        "custom_fields": {}
    },
    "snapshots": {
        "prechange": null,
        "postchange": {
            "created": "2023-08-01T18:51:12.982Z",
            "last_updated": "2023-08-01T18:51:12.982Z",
            "zone": 1,
            "type": "A",
            "name": "ns2.example.org",
            "value": "10.0.0.2",
            "status": "active",
            "ttl": null,
            "managed": false,
            "ptr_record": 7,
            "disable_ptr": false,
            "description": "",
            "ip_address": "10.0.0.2",
            "custom_fields": {},
            "tags": []
        }
    }
}
```

I need the snapshots to figure out what has changed in a record to act accordingly.

The snapshots are generated about here:

https://github.com/netbox-community/netbox/blob/0b10131564dc16138b9b7c7cd869d705771c229e/netbox/extras/webhooks.py#L18-L43

basically a serialize using context={'request':None}

There is no data structure with 
```yaml
nameservers: 
  type: array
  items:
    type: integer
```
in the netbox openapi.yaml description document, so I wonder how this is generated and why the datastructure is missing from the description document.

For the fastapi validation - it uses the models from the netbox openapi.yaml description document.
No problem for the models with netbox I worked with, fails for the netbox-dns models.

Not sure if it helps - but here is the code in question:
In short, it creates pydantic models from the description document using the aiopenapi3 client library, and uses these models to create models for different webhook models-change-messages (IPAddress, Zone, View …) as {name}Event using python generics.

Next, a discriminated union Event using the model property is created - tying the models together.
There is some … challenges with this, parsing the netbox openapi description document is pretty slow, but once the startup is done - enjoy working with the type safe environment provided fastapi/pydantic/aiopenapi3.

```python
import datetime
import uuid
from pathlib import Path
from typing import Annotated, Union, TypeVar, Generic, Literal, ClassVar
import sys

from aiopenapi3 import OpenAPI, FileSystemLoader
from pydantic import BaseModel, Field, RootModel, ConfigDict, create_model


class Error(BaseModel):
    """
    used by HTTPException
    """
    model_config = ConfigDict(allow="forbid")
    detail: str

def createAPI(*names) -> OpenAPI:
    """
    to speed up things, we use some aiopenapi3 plugins to limit the loading process to the schemas required
    removing all paths
    """
    from aiopenapi3.plugin import Init, Document

    class SchemaSelector(Init):
        """
        remove the schemas we do not need models for
        """

        def __init__(self, *schemas):
            super().__init__()
            self.schemas = schemas

        def schema(self, ctx: "Init.Context") -> "Init.Context":
            ctx.schema = {k: ctx.schema[k] for k in (set(self.schemas) & set(ctx.schema.keys()))}
            return ctx

    class RemovePaths(Document):
        def parsed(self, ctx: "Document.Context") -> "Document.Context":
            """
            emtpy the paths - not needed
            """
            ctx.document["paths"] = {}
            return ctx

    # from aiopenapi3.debug import DescriptionDocumentDumper
    # DescriptionDocumentDumper("api-schema.yaml")
    selector = SchemaSelector(*(list(names) + [f"{name}Request" for name in names])) # + [f"Writable{name}Request" for name in names]))
    # dns.yaml = localhost:8000/api/schema/
    api = OpenAPI.load_file("/", Path("dns.yaml"), loader=FileSystemLoader(Path("tests/data")), plugins=[selector, RemovePaths()])
    return api

NAMES = ["IPAddress", "NameServer", "Zone", "View", "Record"]
api = createAPI(*NAMES)

EventTypeT = TypeVar("EventType")
EventTypeRequestT = TypeVar("EventTypeRequest")

class _Snapshots(BaseModel, Generic[EventTypeRequestT]):
    prechange: EventTypeRequestT | None
    postchange: EventTypeRequestT | None


class _Event(BaseModel, Generic[EventTypeT, EventTypeRequestT]):
    """
    generic base class for events
    """
    Snapshot: ClassVar = _Snapshots[EventTypeRequestT]
    """
    required to have the Generic class definition available for use
    will be overwritten in the …Event
    """

    model_config = ConfigDict(extra="forbid")
    event: Literal["created","updated","deleted"]
    timestamp: datetime.datetime
    username: str
    request_id: uuid.UUID
    model: str
    data: EventTypeT
    snapshots: Snapshot[EventTypeRequestT]



def createEvent(name: str, model: str) -> _Event[EventTypeT, EventTypeRequestT]:
    """
    mimic

    class _IPAddressEvent(_Event[api.components.schemas["IPAddress"].get_type()]):
        Snapshot: ClassVar = _Snapshot[api.components.schemas["IPAddress"].get_type()]
        model: Literal["ipaddress"] = "ipaddress"

    :param name: name of class
    :param model: name of netbox model identifier
    """
    Event = api.components.schemas[f"{name}"].get_type()
#    try:
#        EventRequest = api.components.schemas[f"Writable{name}Request"].get_type()
#    except:
    EventRequest = api.components.schemas[f"{name}Request"].get_type()
    m = create_model(f"{name}Event",
                     __base__=(_Event[Event, EventRequest],),
                     model=(Literal[model], model),
                     Snapshot=(ClassVar, _Snapshots[EventRequest]))
    return m


def createEvents(*names):
    events = list()
    for name in names:
        e = createEvent(name, name.lower())
        setattr(sys.modules[__name__], f"{name}Event", e)
        events.append(e)
    return tuple(events)

class Event(RootModel[Annotated[Union[createEvents(*NAMES)], Field(discriminator="model")]]):
    def __getattr__(self, item):
        return getattr(self.root, item)

```

FastAPI uses the defined models.Event structure.

```python
…
@app.post("/webhook", response_model=None, status_code=204, operation_id="webhook",
          responses={"400":{"model": models.Error}})
async def webhook(
    _data: Annotated[models.Event, Body()],
    request: Request,
    response: Response,
    content_length: int = Header(None),
    x_hook_signature: str = Header(None)
) -> None:
```

Don't try to create something from the snippets, wait for me to publish the project,

**Comment by commonism** [comment]
IPAddress update - works

```json
{
    "event": "updated",
    "timestamp": "2023-08-01 19:14:30.806192+00:00",
    "model": "ipaddress",
    "username": "admin",
    "request_id": "816bbf7c-9d3e-4c01-ae7b-4b74115f801b",
    "data": {
        "id": 1,
        "url": "/api/ipam/ip-addresses/1/",
        "display": "10.0.0.1/32",
        "family": {
            "value": 4,
            "label": "IPv4"
        },
        "address": "10.0.0.1/32",
        "vrf": null,
        "tenant": null,
        "status": {
            "value": "active",
            "label": "Active"
        },
        "role": {
            "value": "vrrp",
            "label": "VRRP"
        },
        "assigned_object_type": null,
        "assigned_object_id": null,
        "assigned_object": null,
        "nat_inside": null,
        "nat_outside": [],
        "dns_name": "foffy1234.dx",
        "description": "yes!",
        "comments": "",
        "tags": [],
        "custom_fields": {},
        "created": "2023-07-26T06:02:42.514582Z",
        "last_updated": "2023-08-01T19:14:30.771034Z"
    },
    "snapshots": {
        "prechange": {
            "created": "2023-07-26T06:02:42.514Z",
            "last_updated": "2023-08-01T19:12:58.663Z",
            "description": "yes!",
            "comments": "",
            "address": "10.0.0.1/32",
            "vrf": null,
            "tenant": null,
            "status": "active",
            "role": "vrrp",
            "assigned_object_type": null,
            "assigned_object_id": null,
            "nat_inside": null,
            "dns_name": "foffy1234.dx",
            "custom_fields": {},
            "tags": []
        },
        "postchange": {
            "created": "2023-07-26T06:02:42.514Z",
            "last_updated": "2023-08-01T19:14:30.771Z",
            "description": "yes!",
            "comments": "",
            "address": "10.0.0.1/32",
            "vrf": null,
            "tenant": null,
            "status": "active",
            "role": "vrrp",
            "assigned_object_type": null,
            "assigned_object_id": null,
            "nat_inside": null,
            "dns_name": "foffy1234.dx",
            "custom_fields": {},
            "tags": []
        }
    }
}
```

**Comment by peteeckel** [comment]
Hi @commonism, thanks for the update. Let's have a go at this one-by-one:

1. The 'Zone update' payload confirms what Abhimanyu told me: While the actual data contain the fully deserialised object, the snapshots do not. Since FastAPI is also validating the snapshots this leads to the error you're seeing. This is, Abhimanyu said, probably not a bug but intended behaviour.
2. The second set of errors seems to indicate that there are still some more `allow_null` settings missing, namely for the nested zone and nested record serializer. I've added these now and will push the change shortly. The last error again refers to a snapshot - not sure if we can do something about this, but I'm on it. The third one is more interesting, here it seems that the same issue occurs in the data, not in the snapshot. I will look into it, but I can't promise it will be very short-term.
3. It's interesting to see that the snapshots for the `IPAddress` model seem to be constructed in a different way. Instead of using the IDs of the object, they are using the names as it seems ... `role` in the data is an object, in the snapshot it's a name of a role. That's different from how it's handled for NetBox DNS and I will look at what the differences in the implementation are. Again, no promises regarding the time frame, but it seems to be a trace worth pursuing.

Anyway, I'm looking forward to your project being published - I will certainly build a test instance and test against it so we can get rid of the issues with the API where possible, whatever causes them.

**Comment by peteeckel** [comment]
Hi @commonism, regarding the different handling of `ipam.IPAddress` and `netbox_dns.record.Record` I found the difference. When you look at the serializers for 'family', 'status' and 'role' you'll find that the fields are defined as `ChoiceFields` in the serialiser, not as nested serialisers ... that at least explains that they are handled completely different in the snapshot data.

Unfortunately there are no `ChoiceField`-serialisers in NetBox DNS, so I can't cross-check. But I'm fairly sure that is the reason. You can try that with a VRF or a Tenant ... and what happens is exactly the same as with NetBox DNS:

```
{
    "event": "updated",
    "timestamp": "2023-08-02 16:22:55.497607+00:00",
    "model": "ipaddress",
    "username": "admin",
    "request_id": "0336a91c-ac15-46c3-a61d-90e92464de93",
    "data": {
        "id": 1,
        "url": "/api/ipam/ip-addresses/1/",
        "display": "10.0.0.1/24",
        "family": {
            "value": 4,
            "label": "IPv4"
        },
        "address": "10.0.0.1/24",
        "vrf": {
            "id": 1,
            "url": "/api/ipam/vrfs/1/",
            "display": "Test VRF",
            "name": "Test VRF",
            "rd": null
        },
        "tenant": null,
        "status": {
            "value": "active",
            "label": "Active"
        },
        "role": null,
        "assigned_object_type": null,
        "assigned_object_id": null,
        "assigned_object": null,
        "nat_inside": null,
        "nat_outside": [],
        "dns_name": "test.example.com",
        "description": "",
        "comments": "",
        "tags": [],
        "custom_fields": {},
        "created": "2023-08-02T16:20:08.836426Z",
        "last_updated": "2023-08-02T16:22:55.458295Z"
    },
    "snapshots": {
        "prechange": {
            "created": "2023-08-02T16:20:08.836Z",
            "last_updated": "2023-08-02T16:22:37.745Z",
            "description": "",
            "comments": "",
            "address": "10.0.0.1/24",
            "vrf": 1,
            "tenant": null,
            "status": "active",
            "role": "",
            "assigned_object_type": null,
            "assigned_object_id": null,
            "nat_inside": null,
            "dns_name": "test.example.com",
            "custom_fields": {},
            "tags": []
        },
        "postchange": {
            "created": "2023-08-02T16:20:08.836Z",
            "last_updated": "2023-08-02T16:22:55.458Z",
            "description": "",
            "comments": "",
            "address": "10.0.0.1/24",
            "vrf": 1,
            "tenant": null,
            "status": "active",
            "role": "",
            "assigned_object_type": null,
            "assigned_object_id": null,
            "nat_inside": null,
            "dns_name": "test.example.com",
            "custom_fields": {},
            "tags": []
        }
    }
}
```

You see that in `data` the VRF is exported as a nested object, in `snapshots` only its ID is present. So the behaviour of `IPAddress` is no different from the NetBox DNS models.

That has two consequences: 1. The cause of the issue is most likely within NetBox, 2. As the NetBox core team doesn't regard this a bug it's unlikely that it will be fixed short-term.

**Comment by peteeckel** [comment]
I found the likely explanation for the `address_record` mystery as well.

`address_record` is not a field by itself, but the related name pointing back to records having a `ptr_record` associated with them. For exaple, if `an.example.com` is an A record for the IP address `10.0.0.1`, then `1.0.0.10.in-addr.arpa` has it as its `address_record` because the `ptr_record` of the address records points to it. This is a Django mechanism, not something one has to do oneself.

Using the API, you can now look up the `ptr_record` for an `address_record` and vice versa:
```
    {
      "id": 60,
      "url": "http://192.168.106.105/api/plugins/netbox-dns/records/60/",
      "zone": {
        "id": 1,
        "url": "http://192.168.106.105/api/plugins/netbox-dns/zones/1/",
        "display": "zone1.example.com",
        "name": "zone1.example.com",
        "view": null,
        "status": "active",
        "active": null
      },
      "display": "name01.zone1.example.com [A]",
      "type": "A",
      "name": "name01",
      "value": "10.0.0.1",
      "status": "active",
      "ttl": null,
      "description": "",
      "tags": [],
      "created": "2023-08-01T09:56:30.724449Z",
      "last_updated": "2023-08-02T16:46:50.390624Z",
      "managed": false,
      "disable_ptr": false,
      "ptr_record": {
        "id": 74,
        "url": "http://192.168.106.105/api/plugins/netbox-dns/records/74/",
        "display": "1.0.0.10.in-addr.arpa [PTR]",
        "type": "PTR",
        "name": "1",
        "value": "name01.zone1.example.com.",
        "status": "active",
        "ttl": null,
        "zone": {
          "id": 24,
          "url": "http://192.168.106.105/api/plugins/netbox-dns/zones/24/",
          "display": "0.0.10.in-addr.arpa",
          "name": "0.0.10.in-addr.arpa",
          "view": null,
          "status": "active",
          "active": null
        },
        "active": null
      },
      "address_record": null,
      "active": true,
      "custom_fields": {}
    },
```
```
    {
      "id": 74,
      "url": "http://192.168.106.105/api/plugins/netbox-dns/records/74/",
      "zone": {
        "id": 24,
        "url": "http://192.168.106.105/api/plugins/netbox-dns/zones/24/",
        "display": "0.0.10.in-addr.arpa",
        "name": "0.0.10.in-addr.arpa",
        "view": null,
        "status": "active",
        "active": null
      },
      "display": "1.0.0.10.in-addr.arpa [PTR]",
      "type": "PTR",
      "name": "1",
      "value": "name01.zone1.example.com.",
      "status": "active",
      "ttl": null,
      "description": "",
      "tags": [],
      "created": "2023-08-02T16:46:50.322233Z",
      "last_updated": "2023-08-02T16:46:50.322247Z",
      "managed": true,
      "disable_ptr": false,
      "ptr_record": null,
      "address_record": {
        "id": 60,
        "url": "http://192.168.106.105/api/plugins/netbox-dns/records/60/",
        "display": "name01.zone1.example.com [A]",
        "type": "A",
        "name": "name01",
        "value": "10.0.0.1",
        "status": "active",
        "ttl": null,
        "zone": {
          "id": 1,
          "url": "http://192.168.106.105/api/plugins/netbox-dns/zones/1/",
          "display": "zone1.example.com",
          "name": "zone1.example.com",
          "view": null,
          "status": "active",
          "active": null
        },
        "active": null
      },
      "active": true,
      "custom_fields": {}
    },
```
However, this not being a real field in the model it seems that the mechanism that constructs the snapshots just ignores it, while the schema lists it:
```
        address_record:
          allOf:
          - $ref: '#/components/schemas/NestedRecord'
          readOnly: true
          nullable: true
          description: Address record defining the PTR
```
I'm a bit confused why FastAPI seems to have an issue with it being returned as None in the `data`, as it seems:
```
{'type': 'model_attributes_type', 'loc': ('body', 'record', 'data', 'address_record'), 'msg': 'Input should be a valid dictionary or object to extract fields from', 'input': None, 'url': 'https://errors.pydantic.dev/2.0.3/v/model_attributes_type'}, 
```
I set `allow_none` to `True` for `address_record`, and that also reflects in the schema so it shouldn't stumble across that anymore. Do you have the latest commits of NetBox DNS and the latest schema? If you need a release to upgrade your installation I can prepare one for you.

**Comment by peteeckel** [comment]
And just as I wrote up the last comment there is an update to NetBox with a new Django version ... I just re-rested the webhook issue with NetBox v3.6-beta1:
```
{
    "event": "updated",
    "timestamp": "2023-08-02 17:53:52.779942+00:00",
    "model": "record",
    "username": "admin",
    "request_id": "cce9ddb2-5ae3-439d-bdde-182cb3e9cc00",
    "data": {
        "id": 74,
        "url": "/api/plugins/netbox-dns/records/74/",
        "zone": {
            "id": 24,
            "url": "/api/plugins/netbox-dns/zones/24/",
            "display": "0.0.10.in-addr.arpa",
            "name": "0.0.10.in-addr.arpa",
            "view": null,
            "status": "active",
            "active": null
        },
        "display": "1.0.0.10.in-addr.arpa [PTR]",
        "type": "PTR",
        "name": "1",
        "value": "name01.zone1.example.com.",
        "status": "active",
        "ttl": null,
        "description": "",
        "tags": [],
        "created": "2023-08-02T16:46:50.322233Z",
        "last_updated": "2023-08-02T17:53:52.603939Z",
        "managed": true,
        "disable_ptr": false,
        "ptr_record": null,
        "address_record": {
            "id": 60,
            "url": "/api/plugins/netbox-dns/records/60/",
            "display": "name01.zone1.example.com [A]",
            "type": "A",
            "name": "name01",
            "value": "10.0.0.1",
            "status": "active",
            "ttl": null,
            "zone": {
                "id": 1,
                "url": "/api/plugins/netbox-dns/zones/1/",
                "display": "zone1.example.com",
                "name": "zone1.example.com",
                "view": null,
                "status": "active",
                "active": true
            },
            "active": true
        },
        "custom_fields": {}
    },
    "snapshots": {
        "prechange": null,
        "postchange": {
            "created": "2023-08-02T16:46:50.322Z",
            "last_updated": "2023-08-02T17:53:52.603Z",
            "zone": 24,
            "type": "PTR",
            "name": "1",
            "value": "name01.zone1.example.com.",
            "status": "active",
            "ttl": null,
            "managed": true,
            "ptr_record": null,
            "disable_ptr": false,
            "description": "",
            "ip_address": "10.0.0.1",
            "custom_fields": {},
            "tags": []
        }
    }
}
```
While `address_record` still doesn't show up in the snapshots, at least now it is populated in `data`.

**Comment by commonism** [comment]
https://github.com/commonism/netbox-webhook

Let me know if you get this running.
I'll update the docs accordingly.

**Comment by peteeckel** [comment]
While I haven't yet succeeded in reproducing exactly that error with plain NetBox models, I found another one while trying. I added a webhook for VRF updates, updated my test VRF and got this:
```
{'type': 'missing', 'loc': ('body', 'vrf', 'data', 'ipaddress_count'), 'msg': 'Field required', 'input': {'id': 1, 'url': '/api/ipam/vrfs/1/', 'display': 'Test VRF', 'name': 'Test VRF', 'rd': None, 'tenant': None, 'enforce_unique': True, 'description': '', 'comments': '', 'import_targets': [], 'export_targets': [], 'tags': [], 'custom_fields': {}, 'created': '2023-08-02T16:22:24.871762Z', 'last_updated': '2023-08-03T10:34:03.789901Z'}, 'url': 'https://errors.pydantic.dev/2.1/v/missing'}
{'type': 'missing', 'loc': ('body', 'vrf', 'data', 'prefix_count'), 'msg': 'Field required', 'input': {'id': 1, 'url': '/api/ipam/vrfs/1/', 'display': 'Test VRF', 'name': 'Test VRF', 'rd': None, 'tenant': None, 'enforce_unique': True, 'description': '', 'comments': '', 'import_targets': [], 'export_targets': [], 'tags': [], 'custom_fields': {}, 'created': '2023-08-02T16:22:24.871762Z', 'last_updated': '2023-08-03T10:34:03.789901Z'}, 'url': 'https://errors.pydantic.dev/2.1/v/missing'}
```
This is a similar case to the `address_record` case I was actually interested in: `ipaddress_count` and `prefix_count` are not model fields but annotations, and while they are in the API (and can also be retrieved by API calls), they do not appear in the webhook data or snapshots.

All in all I think we have come to the result of the investigation of this issue and all of its sub-issues: It's not something specific to NetBox DNS, it's definitely related to the general behaviour of NetBox webhooks.

So you are left with two options: Open an issue with NetBox (we have collected quite a lot of material here that you can use), or try to work around these problems in netbox-webhooks. I don't think I can do anything within the limits of NetBox DNS that solves the specific problems related to it as they are only manifestations of the general NetBox behaviour.

Be it as it may: Thanks a lot for raising this issue! I learned a lot while working on it and at least some small fixes in the NetBox DNS API serialisers fell out of the whole process.

**Comment by commonism** [comment]
I already learned the netbox description document does not really match the protocol.
It's basically impossible to use the generated description document to interface the service.

I'll bring this up in netbox.

**Comment by peteeckel** [comment]
> I already learned the netbox description document does not really match the protocol. It's basically impossible to use the generated description document to interface the service.

It generally does (at least most of it, at least I didn't have too many issues recently), but not for webhooks. They seem to have some potential for improvement.

> I'll bring this up in netbox.

That should be the best route to proceed. Is it OK to close this issue?

**Comment by commonism** [comment]
We'll meet again once the requirements of netbox-webhook are met by netbox.
